// unexpected-htmllike-jsx-adapter/src/ReactElementAdapter.js
import React from 'react';
import ObjectAssign from 'object-assign';


function isRawType(value) {
    var type = typeof value;
    return type === 'string' ||
        type === 'number' ||
        type === 'boolean' ||
        type === 'undefined' ||
        value === null;
}

const DefaultOptions = {
    concatTextContent: false
};

function convertValueTypeToString(value) {

    if (typeof value === 'string') { // Common case can be fasttracked
        return value;
    }

    if (value === null || value === undefined) {
        return '';
    }

    return '' + value;
}

function concatenateStringChildren(accum, value) {
    if (isRawType(value) && accum.length &&
        isRawType(accum[accum.length - 1]))
    {
        accum[accum.length - 1] = convertValueTypeToString(accum[accum.length - 1]) + convertValueTypeToString(value);
        return accum;
    }
    accum.push(value);
    return accum;
}

function isValidChild(child) {
    const typeofChild = typeof child;
    return (typeofChild === 'string' ||
        typeofChild === 'number' ||
        typeofChild === 'function' ||
        (typeof child === 'object' && child !== null));
}

class ReactElementAdapter {

    constructor(options) {
        this._options = ObjectAssign({}, DefaultOptions, options);
    }

    getName(element) {
        if (typeof element.type === 'string') {
            return element.type;
        }

        return element.type.displayName || element.type.name || 'no-display-name';
    }

    getAttributes(element) {

        var realProps = {};
        if (element.props) {
            for (var key in element.props) {
                if (key !== 'children') {
                    realProps[key] = element.props[key];
                }
            }
        }

        if (this._options.includeKeyProp && element.key) {
            realProps.key = element.key;
        }

        if (this._options.includeRefProp && element.ref) {
            realProps.ref = element.ref;
        }

        return realProps;
    }

    getChildren(element) {

        var childrenArray = [];

        // This is not using React.Children.forEach / map / toArray because it drops invalid children,
        // which would be fine, but we want to explicitly include the `expect.it()` function as a valid child
        // to enable inline assertions
        // This mirrors the React.Children.forEach logic, as seen at
        // https://github.com/facebook/react/blob/35962a00084382b49d1f9e3bd36612925f360e5b/src/shared/utils/traverseAllChildren.js
        // with the exception that we remove the nulls
        // Basically strings & numbers && elements are allowed (elements classed as objects & functions for simplicity)
        if (Array.isArray(element.props.children)) {
            childrenArray = childrenArray.concat(element.props.children).filter(child => isValidChild(child));
        } else if (isValidChild(element.props.children)) {

            childrenArray = [ element.props.children ];
        }

        if (this._options.convertToString ||
            (this._options.convertMultipleRawToStrings &&
             childrenArray.length > 1)) {
            childrenArray = childrenArray.reduce((agg, child) => {
                if (child !== null && child !== undefined && isRawType(child)) {
                    child = convertValueTypeToString(child);
                }
                agg.push(child);
                return agg;
            }, []);
        }

        if (this._options.concatTextContent) {
            childrenArray = childrenArray.reduce(concatenateStringChildren, []);
        }

        return childrenArray;
    }

    setOptions(newOptions) {

        this._options = ObjectAssign({}, this._options, newOptions);
    }

    getOptions() {
        return this._options;
    }
}


export default ReactElementAdapter;
