// unexpected-react/lib/react-devtools/frontend/SplitPane.js
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */
'use strict';

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; }

var React = require('react');
var Draggable = require('./Draggable');
var assign = require('object-assign');

var SplitPane = (function (_React$Component) {
  function SplitPane(props) {
    _classCallCheck(this, SplitPane);

    _get(Object.getPrototypeOf(SplitPane.prototype), 'constructor', this).call(this, props);
    this.state = {
      width: props.initialWidth };
  }

  _inherits(SplitPane, _React$Component);

  _createClass(SplitPane, [{
    key: 'onMove',
    value: function onMove(x) {
      var node = React.findDOMNode(this);
      this.setState({
        width: node.offsetLeft + node.offsetWidth - x });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this = this;

      var dragStyle = styles.dragger;
      if (this.state.moving) {
        dragStyle = assign({}, dragStyle, styles.draggerMoving);
      }
      var rightStyle = assign({}, styles.rightPane, {
        width: this.state.width });
      return React.createElement(
        'div',
        { style: styles.container },
        React.createElement(
          'div',
          { style: styles.leftPane },
          this.props.left()
        ),
        React.createElement(Draggable, {
          style: dragStyle,
          onStart: function () {
            return _this.setState({ moving: true });
          },
          onMove: function (x) {
            return _this.onMove(x);
          },
          onStop: function () {
            return _this.setState({ moving: false });
          }
        }),
        React.createElement(
          'div',
          { style: rightStyle },
          this.props.right()
        )
      );
    }
  }]);

  return SplitPane;
})(React.Component);

var styles = {
  container: {
    display: 'flex',
    minWidth: 0,
    flex: 1 },

  dragger: {
    cursor: 'ew-resize',
    borderWidth: '0 5px',
    backgroundColor: '#ccc',
    width: 1,
    borderStyle: 'solid',
    borderColor: 'white' },

  draggerMoving: {
    backgroundColor: '#aaf' },

  rightPane: {
    display: 'flex' },

  leftPane: {
    display: 'flex',
    minWidth: 0,
    flex: 1 } };

module.exports = SplitPane;
